### PR TITLE
fix #6840 handle error return values in conda.sh

### LINKED ISSUE #6840: conda install incorrectly returns zero status
**I'm submitting a bug report**

## Current Behavior
`conda install` returns zero status after failing to install a package.

### Steps to Reproduce
```
% conda install no-such-package
Solving environment: failed

PackagesNotFoundError: The following packages are not available from current channels:

  - no-such-package

Current channels:

  - https://repo.continuum.io/pkgs/main/linux-64
  - https://repo.continuum.io/pkgs/main/noarch
  - https://repo.continuum.io/pkgs/free/linux-64
  - https://repo.continuum.io/pkgs/free/noarch
  - https://repo.continuum.io/pkgs/r/linux-64
  - https://repo.continuum.io/pkgs/r/noarch
  - https://repo.continuum.io/pkgs/pro/linux-64
  - https://repo.continuum.io/pkgs/pro/noarch

% echo $?
0
```

## Expected Behavior
Return non-zero status.

## Environment Information
```
     active environment : foo
    active env location : /home/pm/miniconda/envs/foo
            shell level : 1
       user config file : /home/pm/.condarc
 populated config files : 
          conda version : 4.4.8
    conda-build version : 3.4.0
         python version : 3.6.4.final.0
       base environment : /home/pm/miniconda  (writable)
           channel URLs : https://repo.continuum.io/pkgs/main/linux-64
                          https://repo.continuum.io/pkgs/main/noarch
                          https://repo.continuum.io/pkgs/free/linux-64
                          https://repo.continuum.io/pkgs/free/noarch
                          https://repo.continuum.io/pkgs/r/linux-64
                          https://repo.continuum.io/pkgs/r/noarch
                          https://repo.continuum.io/pkgs/pro/linux-64
                          https://repo.continuum.io/pkgs/pro/noarch
          package cache : /home/pm/miniconda/pkgs
                          /home/pm/.conda/pkgs
       envs directories : /home/pm/miniconda/envs
                          /home/pm/.conda/envs
               platform : linux-64
             user-agent : conda/4.4.8 requests/2.18.4 CPython/3.6.4 Linux/4.13.0-32-generic ubuntu/16.04 glibc/2.23
                UID:GID : 1000:1000
             netrc file : None
           offline mode : False
```

### PR DESCRIPTION
**Pull request by kalefranz**
fix #6840

### CODE REVIEW

File: conda/shell/etc/profile.d/conda.sh [hunk 1]
```diff
@@ -84,17 +84,17 @@ conda() {
...
+            _conda_activate "$@" || return $?
             ;;
         deactivate)
-            _conda_deactivate "$@"
+            _conda_deactivate "$@" || return $?
             ;;
         install|update|uninstall|remove)
-            $_CONDA_EXE "$cmd" "$@"
-            _conda_reactivate
+            $_CONDA_EXE "$cmd" "$@" || return $?
```
Comment (mbargull) [inline conda/shell/etc/profile.d/conda.sh]: I think only this line should be changed and I'd rather change it to
```bash
            $_CONDA_EXE "$cmd" "$@" && _conda_reactivate
```
For the `eval` lines a `|| return $?` would be rather arbitrary as only the last evaluated command will affect `||`.
For all `case`s , their return codes should be returned anyway as those are the last evaluated commands (nothing else follows `esac`).
(This also implies that for `(de)activate` the return codes are actually the ones from `_conda_hashr`..)

**Comment by kalefranz** [comment]
Ok, basically a whole new PR.  I added a regression test, which indeed failed for the `conda install` case but passed for the `conda list` case.  The second commit fixes the failure exposed by the regression test.

**Comment by kalefranz** [comment]
And I like this change much better too!

File: tests/test_activate.py [hunk 2]
```diff
@@ -1062,6 +1062,12 @@ def basic_posix(self, shell):
         shell.assert_env_var('?', '0', True)
         # TODO: assert that reactivate worked correctly
 
+        # regression test for #6840
+        shell.sendline('conda install --blah')
+        shell.assert_env_var('?', '2', use_exact=True)
+        shell.sendline('conda list --blah')
+        shell.assert_env_var('?', '2', use_exact=True)
```
Comment (kalefranz) [inline tests/test_activate.py]: I'm not sure if the exit code `2` is really correct. http://tldp.org/LDP/abs/html/exitcodes.html
But it's what argparse does.  From https://github.com/python/cpython/blob/v3.6.4/Lib/argparse.py#L2389 via https://github.com/conda/conda/blob/4.4.9/conda/cli/conda_argparse.py#L147

**Comment by maddenp** [comment]
This issue didn't seem to disappear for me with 4.4.10, so I waited for 4.4.11, but:
```
~ % conda --version
conda 4.4.11
~ % conda install no-such-package
Solving environment: failed

PackagesNotFoundError: The following packages are not available from current channels:

  - no-such-package

Current channels:

  - https://repo.continuum.io/pkgs/main/linux-64
  - https://repo.continuum.io/pkgs/main/noarch
  - https://repo.continuum.io/pkgs/free/linux-64
  - https://repo.continuum.io/pkgs/free/noarch
  - https://repo.continuum.io/pkgs/r/linux-64
  - https://repo.continuum.io/pkgs/r/noarch
  - https://repo.continuum.io/pkgs/pro/linux-64
  - https://repo.continuum.io/pkgs/pro/noarch


~ % echo $?
0
~ % conda info

     active environment : None
            shell level : 0
       user config file : /home/pmadden/.condarc
 populated config files : 
          conda version : 4.4.11
    conda-build version : 3.6.0
         python version : 3.6.4.final.0
       base environment : /media/pmadden/o/miniconda  (writable)
           channel URLs : https://repo.continuum.io/pkgs/main/linux-64
                          https://repo.continuum.io/pkgs/main/noarch
                          https://repo.continuum.io/pkgs/free/linux-64
                          https://repo.continuum.io/pkgs/free/noarch
                          https://repo.continuum.io/pkgs/r/linux-64
                          https://repo.continuum.io/pkgs/r/noarch
                          https://repo.continuum.io/pkgs/pro/linux-64
                          https://repo.continuum.io/pkgs/pro/noarch
          package cache : /media/pmadden/o/miniconda/pkgs
                          /home/pmadden/.conda/pkgs
       envs directories : /media/pmadden/o/miniconda/envs
                          /home/pmadden/.conda/envs
               platform : linux-64
             user-agent : conda/4.4.11 requests/2.18.4 CPython/3.6.4 Linux/4.8.17-galliumos galliumos/2.1 glibc/2.23
                UID:GID : 1000:1000
             netrc file : None
           offline mode : False
```
Is there something else I need to do to get the fix, or is it coming in a future version?

**Pull request merged**